mscore/scoreview.cpp

```diff
@@ -4312,47 +4312,54 @@ static bool elementLower(const Element* e1, const Element* e2)
             return false;
       if (!e2->selectable())
             return true;
+      if (e1->isNote() && e2->isStem())
+            return true;
+      if (e2->isNote() && e1->isStem())
+            return false;
       if (e1->z() == e2->z()) {
+            // same stacking order, prefer non-hidden elements
             if (e1->type() == e2->type()) {
                   if (e1->type() == ElementType::NOTEDOT) {
                         const NoteDot* n1 = static_cast<const NoteDot*>(e1);
                         const NoteDot* n2 = static_cast<const NoteDot*>(e2);
                         if (n1->note() && n1->note()->hidden())
-                              return n2;
-                        else
-                              return n1;
+                              return false;
+                        else if (n2->note() && n2->note()->hidden())
+                              return true;
                         }
                   else if (e1->type() == ElementType::NOTE) {
                         const Note* n1 = static_cast<const Note*>(e1);
                         const Note* n2 = static_cast<const Note*>(e2);
                         if (n1->hidden())
-                              return n2;
-                        else
-                              return n1;
+                              return false;
+                        else if (n2->hidden())
+                              return true;
                         }
                   }
+            // different types, or same type but nothing hidden - use track
+            return e1->track() <= e2->track();
             }
-      return e1->z() < e2->z();
-      }
-
 
+      // default case, use stacking order
+      return e1->z() <= e2->z();
+      }
 
 //---------------------------------------------------------
-//   elementNear
+//   elementsNear
```

Comment:
Can this function be used or modified easily to improve drag-drop behaviour? So that you don't need to put the cursor on the element, but only close to the element for it to consider it as a valid drop target and highlight the element? That could further improve the UX of dragging from the palettes.

Reply:
It could be used for that purpose, but currently is not.  There is elementNear and elementAt, and drag & drop uses the latter.  elementNear honors the proximity setting in Edit / Preferences / Canvas.  In theory we could change drag & drop over to this as well, but I'd worry that by making it less precise, you get all sorts of cases where you dropped on the wrong element.  Consider for example if the proximity zone for a note extended to the articulation, or that of a time signature extended to the key etc.

Reply:
Yes, well, I realise that it will be not as straightforward as just dropping on a random element within x px from the cursor. What I envision, is that drag and drop snaps to the closest valid drop target within a reasonable radius instead of requiring the user to move the mouse precisely to the right place. If there is more than one valid drop target with in our arbitrary radius, we would need to snap to the closest. If two are the same distance, there things could get interesting...

I realise that this could be quite an undertaking, so I am asking whether you think it is possible to do this relatively quickly and easily, and also if you think it is a good idea at all.

Reply:
Well, as I said, I have my doubts, but if you'd like to try, replace all the calls to elementAt or elementsAt in dragdrop.cpp with elementNear and elementsNear (starting from my branch).  It might be trivial and just happen to work, or, like you say, it might open end up being more an undertaking.